core/templates/components/forms/validators/schema-validators.ts

```diff
@@ -0,0 +1,170 @@
+        };
+      }
+      if (!(typeof value === 'string' || Array.isArray(value))) {
+        throw new Error(
+          'Not valid value passed in control. Expecting a string or Array');
+      }
+      if (value.length >= args.minValue) {
+        return null;
+      }
+      return {
+        hasLengthAtLeast: {
+          minValue: args.minValue, actual: control.value.length}
+      };
+    };
+  }
+
+  static hasLengthAtMost(
+      args: {maxValue: number}
+  ): (control: AbstractControl) => ValidationErrors | null {
+    return (control: AbstractControl): ValidationErrors | null => {
+      const value = control.value;
+      if (value === null || value === undefined) {
+        return {
+          hasLengthAtMost: {minValue: args.maxValue, actual: control.value}
+        };
+      }
+      if (!(typeof value === 'string' || Array.isArray(value))) {
+        throw new Error(
+          'Not valid value passed in control. Expecting a string or Array');
+      }
+      if (value.length <= args.maxValue) {
+        return null;
+      }
+      return {
+        hasLengthAtMost: {maxValue: args.maxValue, actual: control.value.length}
+      };
+    };
+  }
+
+  static isAtLeast(
+      args: {minValue: number}
+  ): (control: AbstractControl) => ValidationErrors | null {
+    return (control: AbstractControl): ValidationErrors | null => {
+      const value = parseFloat(control.value);
+      if (!isNaN(value) && value >= args.minValue) {
+        return null;
+      }
+      return {
+        isAtLeast: {minValue: args.minValue, actual: control.value}
+      };
+    };
+  }
+
+  static isAtMost(
+      args: {maxValue: number}
+  ): (control: AbstractControl) => ValidationErrors | null {
+    return (control: AbstractControl): ValidationErrors | null => {
+      const value = parseFloat(control.value);
+      if (!isNaN(value) && value <= args.maxValue) {
+        return null;
+      }
+      return {
+        isAtMost: {maxValue: args.maxValue, actual: control.value}
+      };
+    };
+  }
+
+  static isFloat(): (control: AbstractControl) => ValidationErrors | null {
+    return (control: AbstractControl): ValidationErrors | null => {
+      var FLOAT_REGEXP = /(?=.*\d)^\-?\d*(\.|\,)?\d*\%?$/;
+      // This regex accepts floats in the following formats:
+      // 0.
+      // 0.55..
+      // -0.55..
+      // .555..
+      // -.555..
+      // All examples above with '.' replaced with ',' are also valid.
+      // Expressions containing % are also valid (5.1% etc).
+
+      var viewValue = '';
+      try {
+        viewValue = control.value.toString().trim();
+      } catch (e) {
+        return null;
+      }
+
+      if (viewValue !== '') {
+        if (FLOAT_REGEXP.test(viewValue)) {
+          if (viewValue.slice(-1) === '%') {
+            // This is a percentage, so the input needs to be divided by 100.
+            return isNaN(parseFloat(
+              viewValue.substring(0, viewValue.length - 1).replace(',', '.')
```

Comment:
Please check with @Lawful2002 if this should be dependent on a language service instead.

Reply:
Any number (represented as a string) can be converted to the English format (javascript number type) by using [this function](https://github.com/oppia/oppia/blob/50d8cc64290cab0fe05054d2597a0a1870a8ac56/core/templates/services/number-conversion.service.ts#L57). However, it should be noted that if the input string isn't a valid number, the function would return `null` instead of `NaN`.

Reply:
Same comment as above -- does this need to be fixed in the future? If so, please file an issue and add a TODO linking to it.

@Lawful2002 -- might need your input here, too.

Reply:
I think we filed an issue for this.

Reply:
Please link to that issue and add a TODO if relevant.

Reply:
Done!